### PR TITLE
[WIP] PR #509 merged & with small fixes

### PR DESCRIPTION
**Pull request by gandrewstone**

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -5935,14 +5935,14 @@ bool ProcessMessage(CNode* pfrom, std::string strCommand, CDataStream& vRecv, in
 
     else if (strCommand == NetMsgType::XPEDITEDBLK)
     {
-        // ignore the expedited message unless we are near the chain tip...
-        if (!fImporting && !fReindex && IsChainNearlySyncd())
+        // ignore the expedited message unless we are at the chain tip...
+        if (!fImporting && !fReindex && !IsInitialBlockDownload() && IsThinBlocksEnabled())
```
Comment (kyuupichan) [inline src/main.cpp]: The ThinBlocksEnabled check is now handle by HandleExpeditedBlock after ptschip's recent PR that went in.

File: src/thinblock.cpp [hunk 2]
```diff
@@ -247,6 +248,109 @@ bool CXThinBlock::CheckBlockHeader(const CBlockHeader &block, CValidationState &
...
+/**
+ * Handle an incoming Xthin or Xpedited block
+ * Once the block is validated apart from the Merkle root, forward the Xpedited block with a hop count of nHops.
+ */
+bool CXThinBlock::HandleMessage(CDataStream &vRecv, CNode *pfrom, string strCommand, unsigned nHops)
+{
+    if (!pfrom->ThinBlockCapable())
+    {
+        LOCK(cs_main);
+        Misbehaving(pfrom->GetId(), 100);
```
Comment (kyuupichan) [inline src/thinblock.cpp]: ptschip made this a 5 not 100 in expedited.cpp; we should do the same here

**Review by kyuupichan — APPROVED** [review]

File: src/main.cpp [hunk 3]
```diff
@@ -5936,7 +5936,7 @@ bool ProcessMessage(CNode* pfrom, std::string strCommand, CDataStream& vRecv, in
     else if (strCommand == NetMsgType::XPEDITEDBLK)
     {
         // ignore the expedited message unless we are at the chain tip...
-        if (!fImporting && !fReindex && !IsInitialBlockDownload() && IsThinBlocksEnabled())
```
Comment (ptschip) [inline src/main.cpp]: This is actually not a redundant check and is needed in order to turn thinblocks off completely...so here we're checking if *we* have thinblocks turn off and if so then we don't accept any expedited xthins.  The other check below tell us if the remote peer has it turned on or not.

**Review by ptschip — APPROVED** [review]
ok, never mind my previous comment, looks like that call was moved into expedited....all looks good.

File: src/main.cpp [hunk 4]
```diff
@@ -6076,73 +6076,10 @@ bool ProcessMessage(CNode* pfrom, std::string strCommand, CDataStream& vRecv, in
...
-        if (!pfrom->mapThinBlocksInFlight.count(inv.hash) && !IsExpeditedNode(pfrom))
-        {
-                LOCK(cs_main);
-                Misbehaving(pfrom->GetId(), 100);
-                return error("unrequested thinblock from peer %s (%d)", pfrom->addrName.c_str(), pfrom->id);
-        }
-        }
-
-        thinBlock.process(pfrom, nSizeThinBlock, strCommand);
+        return CXThinBlock::HandleMessage(vRecv, pfrom, strCommand, 0);
```
Comment (gandrewstone) [inline src/main.cpp:6082]: This is deserializing the thin block into an Xthin block, which will not work.

**Comment by gandrewstone** [comment]
PR509 is also attempting to merge the handling of the thin block and the xthin block code.  Although there are similarities like checking headers, there are also significant differences so the code as offered will not work.  I'm going to let @kyuupichan decide where to take this from here.

**Comment by kyuupichan** [comment]
No, it attempting to merge the xthin and expedited code, which are the same thing.  We tested this for a while - 3 of us - and thin, xthin and expedited blocks were all working fine.  What is the issue you're seeing?

**Comment by kyuupichan** [comment]
Ugh, it is incorrect now.  Either my rebase or yours messed this up,  I'll fix and repush

**Comment by sickpig** [comment]
superseded by #525, @gandrewstone should be closed imho. 

edit: @gandrewstone since #525 has already been merged this one need to be closed

**Pull request closed without merging**